inst/resources/lua/custom-environment.lua

```diff
@@ -0,0 +1,197 @@
+    definition = 'def',
+    example = 'exm',
+    exercise = 'exr'
+}
+
+-- other special proof envs
+local proof_label = {
+    proof = 'Proof',
+    remark = 'Remark',
+    solution = 'Solution'
+}
+
+-- for debuging purpose
+local debug_mode = os.getenv("DEBUG_PANDOC_LUA") == "TRUE"
+function print_debug(label,obj,iter)
+    obj = obj or nil
+    iter = iter or pairs
+    label = label or ""
+    label = "DEBUG (from custom-environment.lua): "..label
+    if (debug_mode) then
+        if not obj then
+          print(label)
+        elseif (type(obj) == "string") then
+            print(label.." "..obj)
+        elseif type(obj) == "table" then
+            for k,v in iter(obj) do
+                print(label.."id:"..k.. " val:"..v)
+            end
+        end
+    end
+    return nil
+end
+
+-- create a unique id for a div with none provided
+local counter = 0
+function unlabeled_div()
+    counter = counter + 1 
+    return "unlabeled-div-"..(counter)
+end
+
+-- Get metadata specific to bookdown for this filter
+Meta = function(m) 
+    bookdownmeta = m.bookdown
+    if (bookdownmeta ~= nil) then
+        -- For internationalization feature of bookdown
+        if (bookdownmeta.language ~= nil) then
+            if (bookdownmeta.language.label ~= nil) then
+                for k,v in pairs(bookdownmeta.language.label) do
+                    if (type(v) == 'table' and v.t == 'MetaInlines' and proof_label[k] ~= nil) then
+                        -- remove any undesired space (3 or less)
+                        proof_label[k] = pandoc.utils.stringify(v):gsub("%.?%s?%s?%s?$", "")
+                        print_debug("Translation-> "..k..":", proof_label[k])
+                    end
+                end
+            end
+        end
+    end
+end
+
+-- Modify Pandoc AST for supported custom environment
+Div = function (div)
+    local classes = div.classes
+    -- we do nothing if no classes
+    if (#classes == 0) then
+        print_debug("No classes in the Div.")
+        return div
+    end
+    print_debug("Div classes -> " , classes)
+
+    -- checking if the class is one of the supported custom environment
+    local env_type = {type = nil, env = nil}
+    for i,v in ipairs(classes) do
+        if (theorem_abbr[v] ~= nil) then 
+            env_type.type = "theorem"
+            env_type.env = v
+            break
+        elseif (proof_label[v] ~= nil) then 
+            env_type.type = "proof"
+            env_type.env = v
+            break
+        end
+    end
+    -- classes is not a supported one, we return as is
+    if not env_type.env then
+        print_debug("Not a bookdown supported custom class")
+        return div
+    end
+    print_debug("Found types -> ", env_type)
+
+    -- get the id if it exists - it will we use to build label for reference
+    local id = div.identifier
+    -- if no id, one is generated so that bookdown labelling mechanism works
+    if #id == 0 then id = unlabeled_div() end
```

Comment:
can be simplified as

```lua
local id = div.identifier and unlabeled_div()
```

Reply:
Does it work if `div.identifier` is `""` (empty string) and not `nil` ? 

Because I know I encounter this issue while working on that, and it may be here that I had an issue because otherwise I would have use `if (id ~= nil)` as in other places.

Reply:
Oops, I'm wrong. Sorry.

```lua
> "" and 'foo'
foo
> "bar" and 'foo'
foo
```

Reply:
Yeah that was the tricky thing and why sometime if was safer to test zero-lengh. But here it is because `div.identifier` does not return `nil` as I would have expected. 🤷